File: agent/library/src/main/java/org/matilda/commands/CommunicationCommandRunner.java
package org.matilda.commands;

import org.matilda.commands.listener.CommandResponseListener;
import org.matilda.commands.listener.CommandResponseListeningInstance;

import javax.inject.Inject;
import java.io.IOException;


public class CommunicationCommandRunner implements CommandRunnerInterface {
    @Inject
    CommandSender mCommandSender;

    @Inject
    CommandResponseListener mCommandResponseListener;

    @Inject
    CommandIdGenerator mCommandIdGenerator;

    @Inject
    CommunicationCommandRunner() {}

    public byte[] run(int registryId, int commandType, byte[] parameter) throws IOException, InterruptedException,
            CommandFailedException {
        int commandId = mCommandIdGenerator.generate();
        try (CommandResponseListeningInstance listeningInstance = mCommandResponseListener.listen(commandId)) {
            mCommandSender.send(registryId, commandType, commandId, parameter);
            return listeningInstance.waitForResponse();
        }
    }
}

File: agent/library/src/main/java/org/matilda/commands/CommunicationCommandRunner.kt
package org.matilda.commands

import kotlinx.coroutines.runBlocking
import me.tatarka.inject.annotations.Inject
import org.matilda.commands.listener.CommandResponseListener


class CommunicationCommandRunner
@Inject internal constructor(private val mCommandSender: CommandSender,
                             private val mCommandResponseListener: CommandResponseListener,
                             private val mCommandIdGenerator: CommandIdGenerator) : CommandRunner {
    override fun run(registryId: Int, commandType: Int, parameter: ByteArray): ByteArray {
        val commandId = mCommandIdGenerator.generate()
        try {
            mCommandResponseListener.listen(commandId).use { listeningInstance ->
                mCommandSender.send(registryId, commandType, commandId, parameter)
                return runBlocking { listeningInstance.waitForResponse() }
            }
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
            throw RuntimeException(e)
        }
    }
}
